inference-engine/src/cldnn_engine/cldnn_infer_request.cpp

```diff
@@ -843,101 +834,81 @@ std::map<std::string, InferenceEngineProfileInfo> CLDNNInferRequest::GetPerforma
+    switch (prec) {
+        case Precision::FP32:
+        case Precision::FP16:
+        case Precision::I8:
+        case Precision::U8:
+        case Precision::BOOL:
+        case Precision::I16:
+        case Precision::U16:
+        case Precision::I32:
+        case Precision::I64: {
+            auto impl = getBlobImpl(is_dev_input ?
+                                    remote_ptr :
+                                    reqBlob->as<gpu::ClBlob>());
+            if (!impl->is_allocated()) {
+                IE_THROW() << str_input_not_allocated;
+            }
+            auto inputMem = impl->getMemory();
+
+            if (!is_dev_input) {
+                if (prec == Precision::I16 || prec == Precision::U16) {
+                    // clDNN doesn't support I16 input precision,
+                    // so have to convert input data to fp32 precision
+                    cldnn::mem_lock<float> ptr{ inputMem, stream };
+                    if (prec == Precision::I16) {
+                        copyToFloat<int16_t>(ptr.data(), inputBlob.get());
+                    } else {
+                        copyToFloat<uint16_t>(ptr.data(), inputBlob.get());
+                    }
+                } else {
+                    auto src_lock = inputBlob->cbuffer();
+                    inputMem->copy_from(stream, src_lock.as<const uint8_t*>());
+                }
             }
-            default:
-                IE_THROW() << "Unsupported input precision " << prec;
+            _nw_ptr->set_input_data(internalName, inputMem);
+            break;
         }
-    } else {
-        // Otherwise, we have to attach to user memory and then copy the data.
-        copyInputData(_nw_ptr, inputName, inputLayout, inputBlob);
+        default:
+            IE_THROW() << "Unsupported input precision " << prec;
     }
 }
 
-void CLDNNInferRequest::PrepareInputDyn(const cldnn::primitive_id &inputName, const Blob &inputBlob) {
-    OV_ITT_SCOPED_TASK(itt::domains::CLDNNPlugin, "CLDNNInferRequest::PrepareInputDyn");
-    // now try to get execution results
-    for (unsigned nb = 0; nb < m_graph->GetNetworksCount(); nb++) {
-        unsigned int mask = 1 << nb;
+void CLDNNInferRequest::prepare_output(const cldnn::primitive_id& outputName, Blob::Ptr& outputBlob) {
+    OV_ITT_SCOPED_TASK(itt::domains::CLDNNPlugin, "CLDNNInferRequest::prepare_output");
+    Blob::Ptr reqBlob = _deviceOutputs.at(outputName);
+    cldnn::primitive_id internalName = outputsMap[outputName];
+    auto _nw_ptr = m_graph->GetNetwork();
+    auto remote_ptr = outputBlob->as<gpu::ClBlob>();
+    auto output_blob_ptr = (reqBlob != outputBlob && remote_ptr != nullptr)
+        ? remote_ptr
+        : reqBlob->as<gpu::ClBlob>();
+    auto impl = getBlobImpl(output_blob_ptr);
+    if (!impl->is_allocated()) {
+        IE_THROW() << str_output_not_allocated;
+    }
+    auto outputMem = impl->getMemory();
+    _nw_ptr->set_output_memory(internalName, outputMem, true, true, true);
```

Comment:
I have a quetion just to make sure,  is this happening only once for a stream? it seems updating output memory ptr of cldnn graph,  so I'd like to ensure it for cldnn::loop, which sets body network' memory  w.r.t that of the outer networks memory only at the first execution of the network. If it is changed afterwards, body network cannot be updated with the memroy so that it will behave wrong.

Reply:
As we discussed offline, I had a performance concern of loop memory reset whenever user blob changed.  However, your change will not update the output if the loop output is not direct output right? IF so, such cases are thougth as unusal in the current dlbenchmark networks, so, no problem. However there is a need to reset memory mapping of body network for corner cases where the loop is the last layer.

Reply:
Updated code contains a new updater version which updates all the memory references inside a **loop** primitive body network for the case when **loop** is an output primitive.
New version avoids any buffer allocations, just replaces the pointers. Hence, it doesn't affects performance.

Reply:
LGTM.